### PR TITLE
Fixing views found to be returning a Queryset

### PR DESCRIPTION
**Pull request by tony-sappe**
**Description:**
Some views were returning a Queryset (either at the top-level or buried in the object) which wasn't a "problem" for the existing endpoints but isn't optimal in any case. Introducing Django-CTE surfaced this lingering issue which was a problem for setting a new value in the API cache, effectively braking the cache. This PR replaces the "band-aid" fix which materialized the queryset into Python primitives in limited situations. Now a more robust check is in place and the offending views were modified to no longer return a queryset.

**Technical details:**
- Introduced a new check in the api cache_decorator.py function wrapper to check if a Queryset object is in the response.data field. It will only run in local mode.
- The check function was a generator until an API responses hit the recursion depth limit :-/
- It might be better to embed the check in a new view class but that can be a future exercise if the need presents itself.

**Requirements for PR merge:**

1. [x] Unit & integration tests updated (N/A)
2. [x] API documentation updated (N/A)
3. [x] Necessary PR reviewers:
    - [x] Backend
4. [x] Matview impact assessment completed (N/A)
5. [x] Frontend impact assessment completed (N/A)
6. [x] Data validation completed (N/A)
7. [x] Appropriate Operations ticket(s) created (N/A)
8. [x] Jira Ticket (N/A)

**Area for explaining above N/A when needed:**
```
These changes shouldn't impact observable behaviors
```

### CODE REVIEW

File: usaspending_api/common/cache_decorator.py [hunk 1]
```diff
@@ -1,13 +1,31 @@
...
 
+def contains_queryset(data: Any) -> bool:
+    """Boring non-generator to traverse a complex object and flag any Querysets"""
+    if isinstance(data, QuerySet):
+        return True
+    elif isinstance(data, dict):
+        for v in data.values():
+            if contains_queryset(v):
+                return True
+    elif isinstance(data, list):
```
Comment (kbard) [inline usaspending_api/common/cache_decorator.py:21]: tuples?  sets?  other iterables?

File: usaspending_api/common/cache_decorator.py [hunk 2]
```diff
@@ -30,32 +48,16 @@ def process_cache_response(self, view_instance, view_method, request, args, kwar
...
-                response.data["results"] = list(response.data["results"])
+            # While returning a Queryset is functional most of the time, it isn't
+            # fully supported by Django Rest Framework. This check was inserted
+            # in local mode to catch if a Queryset is being returned by the view
+            # which could cause an exception when setting the cache
+            if settings.IS_LOCAL and response and not response.is_rendered and contains_queryset(response.data):
+                raise RuntimeError(
+                    "Your view is returning a QuerySet.  QuerySets are not really designed to "
+                    "be pickled and can cause caching issues.  Please materialize the QuerySet "
+                    "using a List or some other more primitive data structure."
```
Comment (kbard) [inline usaspending_api/common/cache_decorator.py:59]: "Thank you, have a lovely day, and don't forget to wash your hands regularly!" ... too much?

Comment (tony-sappe) [reply usaspending_api/common/cache_decorator.py:59]: Over the top

Comment (kbard) [reply usaspending_api/common/cache_decorator.py:59]: Aww

File: usaspending_api/common/cache_decorator.py (hunk 1 above)
Comment (tony-sappe) [reply usaspending_api/common/cache_decorator.py]: Good catch. At first the correction caused problems by significantly increasing recursion depth and hitting the limit. Then I realized strings are Iterables and were blowing up the function calls. It's so obvious in hindsight

Comment (kbard) [reply usaspending_api/common/cache_decorator.py]: lol... stupid strings... sorry for causing a headache

**Pull request merged**